### ISSUE TITLE
Hard-to-read output from --help

### ISSUE THREAD

**Issue by quasicomputational** [pcapriotti/optparse-applicative#310]
In dhall-lang/dhall-to-haskell#113, the `--help` output looks like this:

```
Usage: dhall-to-cabal ([<dhall input file>] [--explain] ([--output-dir-cwd DIR]
                      | [--output-dir-input DIR] | [--output-stdout]) |
                      [--print-type TYPE] [--self-contained] |
                      [--print-default TYPE] | [--version])
  Generate Cabal files from Dhall expressions
```

I think that's quite hard to read with how the `--output-*` flags are broken between lines like that, with the `|` at the start of the line, but then the modal flags having their `|` separators at the end of the line. Maybe it's not helping that `dhall-to-cabal` is moderately long name.

Would it be possible to get something more like this, instead?

```
Usage: dhall-to-cabal
         ( [<dhall input file>] [--explain]
             ( [--output-dir-cwd DIR]
             | [--output-dir-input DIR]
             | [--output-stdout]
             )
         | [--print-type TYPE] [--self-contained]
         | [--print-default TYPE]
         | [--version]
         )
  Generate Cabal files from Dhall expressions
```

**Comment by HuwCampbell** [comment]
I don't disagree.

Pretty printing like this is actually quite a hard problem (and quite subjective), requiring non-greedy search across the possible line breaks, judging the "attractiveness" of the line break combinations.

The current pretty printer doesn't support this, and AFAIK, prettyprinter (which we're likely to move to soon) doesn't either.

**Comment by quasicomputational** [comment]
Yeah, pretty-printing is actually quite a hard problem. Maybe there could be an option to go to a simpler algorithm (every option gets its own line)? Then, if the maintainer of the application decides the smart algorithm isn't producing good results, there's a reasonable fallback.

**Comment by HuwCampbell** [comment]
While #354 didn't quite go this far, I would be interested to see how it might have improved your use case.

I've delved quite a bit into the internals of the pretty printer now, so might also be able to nut out something a bit more along the lines of what you've written down the road.

**Comment by quasicomputational** [comment]
Here's the present `--help` output with optparse-applicative 0.15.0.0:

```
Usage: dhall-to-cabal [[<dhall input file>] [--explain] [--output-stdout |
                      --output-dir-cwd DIR | --output-dir-input DIR] |
                      --print-type TYPE [--self-contained] |
                      --print-default TYPE | --version]
```

Consistently putting the `|` at the ends of lines is definitely an improvement in legibility.

I'd say that the remaining problem is that there are four modes of operation (normal operation, `--print-type`, `--print-default` and `--version`), but the line breaking (and lack of indentation hinting) is making that hard to recognise at a glance, especially because of the three-way `--output-*` flags.

**Comment by HuwCampbell** [comment]
Can you try with current master?

**Comment by HuwCampbell** [comment]
With an additional small change, I'm now getting this
```
Usage: <interactive> [[<dhall input file>] [--explain] 
                       [--output-stdout | --output-dir-cwd DIR |
                         --output-dir-input DIR] | --print-type TYPE 
                       [--self-contained] | --print-default TYPE | --version]
  Generate Cabal files from Dhall expressions
```

Which looks nicer IMO.

**Comment by HuwCampbell** [comment]
See #355

**Comment by quasicomputational** [comment]
That's a nice step towards what I want - the indentation is a big help especially.

I think it could be further improved with some additional linebreaks after `|`. For example, on the third line before `--print-type`, to make it clear that that's an entirely different mode of operation and that the `--self-contained` isn't a flag for the normal mode of operation, which is obscured by how `--print-type` is hiding at the end of a line.

There might be a simple rule here, something like 'if the previous term being pretty-printed has been broken across multiple lines, add a linebreak after the next `|`'.

**Comment by HuwCampbell** [comment]
Yeah, I can see how that would help.
With an additional small change I can get this:

I very much want to balance readability with number of lines,
and don't really think any more spacing out would be a good idea.

```
Usage: <interactive> [[<dhall input file>] [--explain]
                       [--output-stdout | --output-dir-cwd DIR |
                         --output-dir-input DIR] |
                       --print-type TYPE [--self-contained] |
                       --print-default TYPE | --version]
  Generate Cabal files from Dhall expressions
```
